### PR TITLE
Remove time and frequency constraint when calculating PSD

### PR DESCRIPTION
**Pull request by hoechenberger**
Calculation takes longer, but visualization will provide a less distorted view of the power spectrum.

### CODE REVIEW

File: 02-frequency_filter.py [hunk 1]
```diff
@@ -92,8 +92,7 @@ def run_filter(subject, run=None, session=None):
         raw.plot(n_channels=50, butterfly=True)
 
         # plot power spectral densitiy
-        raw.plot_psd(area_mode='range', tmin=10.0, tmax=100.0,
-                     fmin=0., fmax=50., average=True)
+        raw.plot_psd(average=True, area_mode='range', n_fft=len(raw.times))
```
Comment (agramfort) [inline 02-frequency_filter.py:95]: this is not the right fix. You DO NOT want to compute the FFT of the full signal but rather average more windows in the Welch method so do not specify tmin and tmax

I would do:

`raw.plot_psd(area_mode='range', fmin=0., fmax=50., average=False)`

Comment (hoechenberger) [reply 02-frequency_filter.py:95]: Ah you see, I was struggling with the meaning of those parameters :) So what I actually want to change are `nperseg` and `noverlap` passed to `scipy.signal.welch()`, right?
https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.signal.welch.html

Comment (agramfort) [reply 02-frequency_filter.py]: yes but these parameters are not exposed as is in plot_psd

File: 02-frequency_filter.py [hunk 2]
```diff
@@ -92,8 +92,7 @@ def run_filter(subject, run=None, session=None):
         raw.plot(n_channels=50, butterfly=True)
 
         # plot power spectral densitiy
-        raw.plot_psd(area_mode='range', tmin=10.0, tmax=100.0,
-                     fmin=0., fmax=50., average=True)
+        raw.plot_psd(average=True, area_mode='range')
```
Comment (agramfort) [inline 02-frequency_filter.py:95]: I don't think it's a good idea to average. non-averaged plots are excellent ways to spot remaining bad channels

Comment (hoechenberger) [reply 02-frequency_filter.py:95]: Why not plot both? :)

Comment (hoechenberger) [reply 02-frequency_filter.py:95]: Also… the non-avg plot will take quite a while to be rendered :(

**Comment by agramfort** [comment]
My first implementation was using average and then average option was added.
I never see now a need to use average=True but it's there for historical
reasons

>

**Comment by agramfort** [comment]
really? it's just 300 lines so I doubt it's a bottleneck

**Comment by hoechenberger** [comment]
> really? it's just 300 lines so I doubt it's a bottleneck

The bottleneck is matplotlib, when we ask it to plot PSDs for 300+ channels :)
Also the average plot is what @dnacombo uses when looking at empty-room recordings…

**Comment by agramfort** [comment]
I am sure I can convince @dnacombo to do otherwise :)

**Comment by jasmainak** [comment]
are you sure the bottleneck is matplotlib and not the fact that you now compute for the entire recording?

btw, for spotting bad channels you must use log-log plot. Maybe having both plot is useful for a user.

File: 02-frequency_filter.py [hunk 3]
```diff
@@ -92,8 +92,7 @@ def run_filter(subject, run=None, session=None):
         raw.plot(n_channels=50, butterfly=True)
```
Comment (jasmainak) [inline 02-frequency_filter.py:92]: Is `n_jobs = 1` necessarily when `config.plot=True`?

Comment (hoechenberger) [reply 02-frequency_filter.py:92]: Very good question, and it would apply to the entire pipeline!

File: 02-frequency_filter.py (hunk 2 above)
Comment (agramfort) [reply 02-frequency_filter.py:95]: Can you use average False here please ?

**Comment by hoechenberger** [comment]
> btw, for spotting bad channels you must use log-log plot.

Why's that? PSD is in dB already; which should I also log-transform the abscissa (frequencies)?

**Comment by agramfort** [comment]
This suggests using semilogx 

Try on mne sample that has 2 bads to convince yourself

**Comment by jasmainak** [comment]
> Why's that? PSD is in dB already; why should I also log-transform the abscissa (frequencies)?

read our [biomag demo paper](https://www.frontiersin.org/articles/10.3389/fnins.2018.00530/full) Figure 2 :) Many of these bad channels have slow artifacts so you see them easier if the x axis is also in log scale

**Comment by agramfort** [comment]
+10 !

**Comment by hoechenberger** [comment]
Interesting, I would have assumed that it’s mostly hf artifacts. Will put this on my to-read stack :)

**Comment by hoechenberger** [comment]
Closing. We should change the x scaling to logarithmic though.

**Pull request closed without merging**